### PR TITLE
[Merged by Bors] - feat(ConditionallyCompleteLattice): ciSup_image, ciSup_or', and other iSup-like lemmas

### PR DESCRIPTION
**Pull request by pechersky**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/ConditionallyCompleteLattice/Basic.lean [hunk 1]
```diff
@@ -1126,6 +1215,14 @@ theorem ciSup_mono' {ι'} {f : ι → α} {g : ι' → α} (hg : BddAbove (range
 theorem csInf_le_csInf' {s t : Set α} (h₁ : t.Nonempty) (h₂ : t ⊆ s) : sInf s ≤ sInf t :=
   csInf_le_csInf (OrderBot.bddBelow s) h₁ h₂
 
+lemma ciSup_or' (p q : Prop) (x : ι) (f : ι → α) :
+    ⨆ (_ : p ∨ q), f x = (⨆ (_ : p), f x) ⊔ ⨆ (_ : q), f x := by
```
Comment (eric-wieser) [inline Mathlib/Order/ConditionallyCompleteLattice/Basic.lean:1219]:
```suggestion
lemma ciSup_or' (p q : Prop) (f : p ∨ q → α) :
    ⨆ (h : p ∨ q), f h = (⨆ h : p, f (.inl h)) ⊔ ⨆ h : q, f (.inr h) := by
```

File: Mathlib/Order/ConditionallyCompleteLattice/Basic.lean [hunk 2]
```diff
@@ -864,6 +864,95 @@ lemma sup_eq_top_of_top_mem [OrderTop α] (h : ⊤ ∈ s) : sSup s = ⊤ :=
...
+    refine hg'.trans ?_
+    have : Nonempty s := Set.Nonempty.to_subtype hs
+    refine ciSup_le ?_
+    intro ⟨i, h⟩
+    obtain ⟨t, ht⟩ : ∃ t : f '' s, g t = g (f (Subtype.mk i h)) := by
+      have : f i ∈ f '' s := Set.mem_image_of_mem _ h
+      exact ⟨⟨f i, this⟩, by simp [this]⟩
+    rw [← ht]
+    refine le_ciSup_set ?_ t.prop
+    · simpa [bddAbove_def] using hf
```
Comment (Ruben-VandeVelde) [inline Mathlib/Order/ConditionallyCompleteLattice/Basic.lean:946]:
```suggestion
    simpa [bddAbove_def] using hf
```

File: Mathlib/Order/ConditionallyCompleteLattice/Basic.lean [hunk 3]
```diff
@@ -864,6 +864,95 @@ lemma sup_eq_top_of_top_mem [OrderTop α] (h : ⊤ ∈ s) : sSup s = ⊤ :=
...
+    rw [this]
+    refine le_ciSup (f := (fun i : ι ↦ ⨆ (h : p i), f ⟨i, h⟩)) ?_ i
+    simp_rw [ciSup_eq_ite]
+    refine (hf.union (bddAbove_singleton (a := sSup ∅))).mono ?_
+    intro
+    simp only [Set.mem_range, Set.union_singleton, Set.mem_insert_iff, Subtype.exists,
+      forall_exists_index]
+    intro b hb
+    split_ifs at hb
+    · refine Or.inr ⟨_, _, hb⟩
```
Comment (Ruben-VandeVelde) [inline Mathlib/Order/ConditionallyCompleteLattice/Basic.lean:883]:
```suggestion
    · exact Or.inr ⟨_, _, hb⟩
```

**Review by Ruben-VandeVelde — APPROVED** [review]
maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**